File: tools/release/regenerate-changelog.mjs
#!/usr/bin/env node
// Regenerates the root CHANGELOG.md from the repository git history.
//
// Uses `git-cliff` configured by `cliff.toml`. If git-cliff is not on PATH,
// this script downloads the pinned release for the current platform into a
// scratch directory and runs it from there, so the regeneration is
// reproducible on contributor machines and in CI without a global install.

import { spawnSync } from "node:child_process";
import { existsSync, mkdtempSync } from "node:fs";
import { mkdir, chmod } from "node:fs/promises";
import { tmpdir, arch as osArch, platform as osPlatform } from "node:os";
import path from "node:path";
import { fileURLToPath } from "node:url";
import https from "node:https";
import { createWriteStream } from "node:fs";
import { pipeline } from "node:stream/promises";
import { execFile } from "node:child_process";
import { promisify } from "node:util";

const execFileP = promisify(execFile);

const GIT_CLIFF_VERSION = "2.6.1";
const scriptPath = fileURLToPath(import.meta.url);
const __dirname = path.dirname(scriptPath);
const repoRoot = path.resolve(__dirname, "../..");

function resolveAsset() {
  const platform = osPlatform();
  const arch = osArch();
  if (platform === "linux" && arch === "x64") {
    return `git-cliff-${GIT_CLIFF_VERSION}-x86_64-unknown-linux-gnu.tar.gz`;
  }
  if (platform === "linux" && arch === "arm64") {
    return `git-cliff-${GIT_CLIFF_VERSION}-aarch64-unknown-linux-gnu.tar.gz`;
  }
  if (platform === "darwin" && arch === "arm64") {
    return `git-cliff-${GIT_CLIFF_VERSION}-aarch64-apple-darwin.tar.gz`;
  }
  if (platform === "darwin" && arch === "x64") {
    return `git-cliff-${GIT_CLIFF_VERSION}-x86_64-apple-darwin.tar.gz`;
  }
  if (platform === "win32" && arch === "x64") {
    return `git-cliff-${GIT_CLIFF_VERSION}-x86_64-pc-windows-msvc.zip`;
  }
  throw new Error(`Unsupported platform/arch for git-cliff fallback: ${platform}/${arch}`);
}

function hasGitCliffOnPath() {
  const probe = spawnSync("git-cliff", ["--version"], { stdio: "ignore" });
  return probe.status === 0;
}

async function downloadFile(url, dest) {
  await new Promise((resolve, reject) => {
    const request = https.get(url, (response) => {
      if (
        (response.statusCode === 301 ||
          response.statusCode === 302 ||
          response.statusCode === 307 ||
          response.statusCode === 308) &&
        response.headers.location
      ) {
        downloadFile(response.headers.location, dest).then(resolve, reject);
        return;
      }
      if (response.statusCode !== 200) {
        reject(new Error(`HTTP ${response.statusCode} for ${url}`));
        return;
      }
      pipeline(response, createWriteStream(dest)).then(resolve, reject);
    });
    request.on("error", reject);
  });
}

async function ensureGitCliff() {
  if (hasGitCliffOnPath()) return "git-cliff";

  const asset = resolveAsset();
  const url = `https://github.com/orhun/git-cliff/releases/download/v${GIT_CLIFF_VERSION}/${asset}`;
  const scratch = mkdtempSync(path.join(tmpdir(), "vize-git-cliff-"));
  const archive = path.join(scratch, asset);
  await downloadFile(url, archive);

  if (asset.endsWith(".tar.gz")) {
    await execFileP("tar", ["-xzf", archive, "-C", scratch]);
  } else {
    await execFileP("unzip", ["-q", archive, "-d", scratch]);
  }

  const expected = asset.replace(/\.tar\.gz$|\.zip$/, "");
  const bin = path.join(
    scratch,
    expected,
    osPlatform() === "win32" ? "git-cliff.exe" : "git-cliff",
  );
  await chmod(bin, 0o755);
  return bin;
}

export function buildGitCliffArgs(argv) {
  const args = ["--config", "cliff.toml", "--output", "CHANGELOG.md"];

  for (let i = 0; i < argv.length; i++) {
    const arg = argv[i];
    if (arg === "--unreleased") {
      args.push("--unreleased");
      continue;
    }
    if (arg === "--latest") {
      args.push("--latest");
      continue;
    }
    if (arg === "--tag") {
      const tag = argv[i + 1];
      if (tag == null || tag === "" || tag.startsWith("--")) {
        throw new Error("Missing value for --tag");
      }
      args.push("--tag", tag);
      i += 1;
      continue;
    }
    throw new Error(`Unknown argument: ${arg}`);
  }

  return args;
}

async function main() {
  const args = buildGitCliffArgs(process.argv.slice(2));
  await mkdir(path.dirname(path.join(repoRoot, "CHANGELOG.md")), { recursive: true });
  const cliffBin = await ensureGitCliff();

  const result = spawnSync(cliffBin, args, { cwd: repoRoot, stdio: "inherit" });
  if (result.status !== 0) {
    process.exit(result.status ?? 1);
  }

  if (!existsSync(path.join(repoRoot, "CHANGELOG.md"))) {
    console.error("git-cliff finished but CHANGELOG.md is missing.");
    process.exit(1);
  }
}

if (process.argv[1] != null && path.resolve(process.argv[1]) === scriptPath) {
  main().catch((error) => {
    console.error(error);
    process.exit(1);
  });
}

File: tools/commands/release/regenerate-changelog.rs
#!/usr/bin/env rust-script
//! ```cargo
//! [dependencies]
//! serde = { version = "1", features = ["derive"] }
//! serde_json = "1"
//! tempfile = "3"
//!
//! [package]
//! edition = "2024"
//! ```

#[path = "../../support/common.rs"]
mod common;

use std::{
    env, fs,
    path::{Path, PathBuf},
    process::{Command, ExitCode, Stdio},
};

const GIT_CLIFF_VERSION: &str = "2.6.1";

fn main() -> ExitCode {
    common::main_result(run())
}

fn run() -> Result<(), String> {
    let raw_args = env::args().skip(1).collect::<Vec<_>>();
    let (print_args, args) = parse_git_cliff_args(&raw_args)?;
    if print_args {
        println!(
            "{}",
            serde_json::to_string(&args).map_err(|error| error.to_string())?
        );
        return Ok(());
    }

    let root = common::repo_root()?;
    let cliff = ensure_git_cliff()?;
    let status = Command::new(&cliff)
        .args(&args)
        .current_dir(&root)
        .stdin(Stdio::null())
        .status()
        .map_err(|error| format!("failed to run {}: {error}", cliff.display()))?;
    if !status.success() {
        return Err(format!(
            "git-cliff failed with exit {}",
            status.code().unwrap_or(1)
        ));
    }
    let changelog = root.join("CHANGELOG.md");
    if !changelog.is_file() {
        return Err("git-cliff finished but CHANGELOG.md is missing.".to_string());
    }
    Ok(())
}

fn parse_git_cliff_args(argv: &[String]) -> Result<(bool, Vec<String>), String> {
    let mut print_args = false;
    let mut args = vec![
        "--config".to_string(),
        "cliff.toml".to_string(),
        "--output".to_string(),
        "CHANGELOG.md".to_string(),
    ];
    let mut index = 0usize;
    while index < argv.len() {
        match argv[index].as_str() {
            "--print-args" => {
                print_args = true;
                index += 1;
            }
            "--unreleased" | "--latest" => {
                args.push(argv[index].clone());
                index += 1;
            }
            "--tag" => {
                let Some(tag) = argv.get(index + 1) else {
                    return Err("Missing value for --tag".to_string());
                };
                if tag.is_empty() || tag.starts_with("--") {
                    return Err("Missing value for --tag".to_string());
                }
                args.push("--tag".to_string());
                args.push(tag.clone());
                index += 2;
            }
            arg => return Err(format!("Unknown argument: {arg}")),
        }
    }
    Ok((print_args, args))
}

fn ensure_git_cliff() -> Result<PathBuf, String> {
    if Command::new("git-cliff")
        .arg("--version")
        .stdin(Stdio::null())
        .stdout(Stdio::null())
        .stderr(Stdio::null())
        .status()
        .map(|status| status.success())
        .unwrap_or(false)
    {
        return Ok(PathBuf::from("git-cliff"));
    }

    let asset = resolve_asset()?;
    let url = format!(
        "https://github.com/orhun/git-cliff/releases/download/v{GIT_CLIFF_VERSION}/{asset}"
    );
    let scratch = tempfile::Builder::new()
        .prefix("vize-git-cliff-")
        .tempdir()
        .map_err(|error| error.to_string())?
        .keep();
    let archive = scratch.join(&asset);
    download_file(&url, &archive)?;

    let archive_text = archive.to_string_lossy().into_owned();
    let scratch_text = scratch.to_string_lossy().into_owned();
    if asset.ends_with(".tar.gz") {
        common::run_capture(
            "tar",
            &["-xzf", archive_text.as_str(), "-C", scratch_text.as_str()],
        )?;
    } else {
        common::run_capture(
            "unzip",
            &["-q", archive_text.as_str(), "-d", scratch_text.as_str()],
        )?;
    }

    let stem = asset
        .strip_suffix(".tar.gz")
        .or_else(|| asset.strip_suffix(".zip"))
        .unwrap_or(&asset);
    let bin = scratch.join(stem).join(if env::consts::OS == "windows" {
        "git-cliff.exe"
    } else {
        "git-cliff"
    });
    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        fs::set_permissions(&bin, fs::Permissions::from_mode(0o755))
            .map_err(|error| format!("cannot chmod {}: {error}", bin.display()))?;
    }
    Ok(bin)
}

fn resolve_asset() -> Result<String, String> {
    match (env::consts::OS, env::consts::ARCH) {
        ("linux", "x86_64") => Ok(format!(
            "git-cliff-{GIT_CLIFF_VERSION}-x86_64-unknown-linux-gnu.tar.gz"
        )),
        ("linux", "aarch64") => Ok(format!(
            "git-cliff-{GIT_CLIFF_VERSION}-aarch64-unknown-linux-gnu.tar.gz"
        )),
        ("macos", "aarch64") => Ok(format!(
            "git-cliff-{GIT_CLIFF_VERSION}-aarch64-apple-darwin.tar.gz"
        )),
        ("macos", "x86_64") => Ok(format!(
            "git-cliff-{GIT_CLIFF_VERSION}-x86_64-apple-darwin.tar.gz"
        )),
        ("windows", "x86_64") => Ok(format!(
            "git-cliff-{GIT_CLIFF_VERSION}-x86_64-pc-windows-msvc.zip"
        )),
        (platform, arch) => Err(format!(
            "Unsupported platform/arch for git-cliff fallback: {platform}/{arch}"
        )),
    }
}

fn download_file(url: &str, dest: &Path) -> Result<(), String> {
    let output = Command::new("curl")
        .args([
            "--fail",
            "--location",
            "--silent",
            "--show-error",
            "--output",
        ])
        .arg(dest)
        .arg(url)
        .stdin(Stdio::null())
        .output()
        .map_err(|error| format!("failed to run curl: {error}"))?;
    if !output.status.success() {
        return Err(format!(
            "curl failed for {url}\n{}{}",
            String::from_utf8_lossy(&output.stderr),
            String::from_utf8_lossy(&output.stdout)
        )
        .trim()
        .to_string());
    }
    Ok(())
}
